### ISSUE TITLE
[CockroachDB] No difference between `integer` and `bigInteger`

### ISSUE THREAD

**Issue by rijkvanzanten** [knex/knex#4953]
# Environment

Knex version: 0.95.15+
Database + version: CockroachDB v21.1.13
OS: macOS Monterrey

# Bug

Both `integer` and `bigInteger` both result in a `int8` column being created in CockroachDB, whereas Postgres creates them as `int4` and `int8` respectively. This becomes an issue, as `int8`s are always returned as a String by the `pg` connection library, causing unexpected incompatibilities between the database vendors.

```js
knex.schema.createTable('example', (table) => {
  table.increments();

  table.integer('example_int');
  table.bigInteger('example_bigint');
});
```

Note how both columns are created as an `int8` in CockroachDB.

I believe this is due to the fact that CockroachDB defaults to `int8` when using `.. ADD COLUMN example INTEGER`, which is [the default from here](https://github.com/knex/knex/blob/63980987a6f50529f37b96c127c05e6d87abc7d6/lib/schema/columncompiler.js#L282-L285) (AFAICT as the PG dialect doesn't have a specific `integer` in the columnbuilder).

**Comment by kibertoad** [comment]
Would you consider sending a fix for this? Should be fairly straightforward to override this for Cockroach

**Comment by rijkvanzanten** [comment]
Sure thing!

**Comment by rijkvanzanten** [comment]
Are there any current tests I can enhance to check against the type creation @kibertoad @OlivierCavadenti? I noticed the schema specs under integration2 here:

https://github.com/knex/knex/blob/master/test/integration2/schema/index.spec.js

but I wasn't able to spot any test that confirms that a `.integer()` actually creates an integer. 

Alternatively, we might also be able to catch this by adding some check against integer fields created by knex under `integration2/query/select` 🤔

**Comment by kibertoad** [comment]
@rijkvanzanten I don't think we have tests to validate integers being integers. Probably the closes thing would be unit tests which validate SQL being generated

**Comment by rijkvanzanten** [comment]
@kibertoad I added a proof of concept "type consistency" spec file to the integration2 tests. Is that useful to have, or a bit overkill? (https://github.com/knex/knex/blob/9ea1ae98e3d8f4ce38a10971e25768b68105b8d0/test/integration2/schema/type-consistency.spec.js)

**Comment by kibertoad** [comment]
@rijkvanzanten It looks great!

**Comment by OlivierCavadenti** [comment]
> @kibertoad I added a proof of concept "type consistency" spec file to the integration2 tests. Is that useful to have, or a bit overkill? (https://github.com/knex/knex/blob/9ea1ae98e3d8f4ce38a10971e25768b68105b8d0/test/integration2/schema/type-consistency.spec.js)

Quite useful, for example on work binary uuid I did !

**Comment by rijkvanzanten** [comment]
The same applies to `.increments()` and `.bigIncrements()`

**Comment by rijkvanzanten** [comment]
It looks like `.increments()` currently creates a `int8` with default value `unique_rowid()`, which works very differently from Postgres' `int4` with `nextval()` on a sequence. This matches the default behavior as per [the CockroachDB docs](https://www.cockroachlabs.com/docs/stable/serial.html).

If we'd want to make this consistent across databases, I guess we should have `.increments()` create a `int4`, create a sequence, and rely on `nextval()` instead of the current default behavior. 

It looks like CockroachDB has a session flag that might be able to do this automatically:

<img width="1129" alt="CleanShot 2022-01-20 at 18 56 32@2x" src="https://user-images.githubusercontent.com/9141017/150440966-ef02f769-edd4-436f-a860-dbda2a893b96.png">

Which makes we wonder if this is something Knex should handle, or if that's something we should point out in the docs and have the user configure it properly.

I think this comes back to the discussion we've had a couple times before: How much _abstraction_ should knex have to align the different database vendors. We already implemented a bunch of extra custom logic to allow SQLite to play nice with the others, but in this case, a simple set of SET statements in the pool connection callback would solve it. However, in that case the default behavior for CRDB is going to be pretty confusing / unexpected 🤔 There's no point in having `integer` vs `bigInteger` (and their increments counterparts) if they both do the same thing..

Thoughts @kibertoad @OlivierCavadenti?

**Comment by rijkvanzanten** [comment]
Seems like the following configuration should align Cockroaches behavior with how you would expect knex to behave by default:

```js
const knex = require('knex')({
  client: 'pg',
  connection: {...},
  pool: {
    afterCreate: function (conn, done) {
      conn.query('SET serial_normalization = "sql_sequence";', function (err) {
        if (err) {
          done(err, conn);
        } else {
          conn.query('SET default_int_size = 4;', function (err) {
            done(err, conn);
          });
        }
      });
    }
  }
});
```

**Comment by kibertoad** [comment]
This makes me question whether we should be doing any overrides of default CRDB behaviour to begin with. Forcing it to behave like pg sounds too opinionated for a query builder.
And if we keep increments an int8, then integer probably should stay one too, so that it is at least consistent with itself.
What are the implications of current misalignment between pg and crdb? numeric values need to be handled differently? wonder if it should be handled via documentation instead.

**Comment by rijkvanzanten** [comment]
> What are the implications of current misalignment between pg and crdb? numeric values need to be handled differently? 

As of right now, it means that (for CRDB):

- There's no difference between `integer`/`bigInteger` and `increments`/`bigIncrements`
- `.increments` doesn't actually use auto-incrementing integers, like the name suggests
- All integers are returned as Strings at all times

**Comment by nickrum** [comment]
I agree with @rijkvanzanten! Maybe we could add a CRDB-only "postgres compat" option that sets those options and essentially makes it behave more like pg and the other vendors? The question then becomes if this should be enabled by default.

**Comment by kibertoad** [comment]
@rafiss Maybe you could comment on this issue and advise on whether preserving status quo or adding overrides would be better from CockroachDB perspective?

**Comment by kibertoad** [comment]
@nickrum I like that idea, but I wonder how intrusive that would be. Maybe it would be easier to achieve not as an option, but as a separate dialect, extending vanilla cockroachdb one. I would still like to hear @rafiss perspective first, because maybe we would be sabotaging crdb by overriding defaults that might be there for a good reason.

**Comment by rijkvanzanten** [comment]
I'm personally worried that the API of knex itself becomes too unexpected/unreliable between dialects 🤔 There's currently already quite a lot of "It works like X, except for in VENDOR where it does Y".

We could also add a new `.uniqueID()` (name TDB) method that would allow to explicitly use Cockroach's default row identifier, which allows knex to keep a consistent set of methods for the above mentioned problems? For example:

- `.integer()` would use `int4`
- `.bigInteger()` would use `int8`
- `.increments()` would use `serial_normalization = "sql_sequence"` with `int4`
- `.bigIncrements()` would use `serial_normalization = "sql_sequence"` with `int8`
- `.uniqueID()` Use Cockroach's `unique_rowid` based `int8` serial primary key

**Comment by wodka** [comment]
Hi - what would also be possible is to force for integer to create an INT4 - https://www.cockroachlabs.com/docs/stable/int.html
This could be done by setting the session variable to use int4 for integer

**Comment by mercmobily** [comment]
CockroachDB defaults `INTEGER` to 64‑bit, which makes `integer()` and `bigInteger()` indistinguishable when using the Postgres dialect defaults.

Code context:

- lib/schema/columncompiler.js: general default for `integer` flows to the dialect; there’s no Cockroach‑specific override, so the DDL ends up as `INTEGER`, which Cockroach maps to `INT8`.
- The Postgres dialect maps `bigInteger` → `bigint` already.

Suggested fix:

- Add a CockroachDB‑specific column compiler override that maps `integer()` to `int4` (`INT`) and `bigInteger()` to `int8` (`BIGINT`). This preserves behaviour parity with Postgres and avoids the `pg` client returning strings for `INT8` where not intended.
- Integration tests like your proposed `type-consistency.spec.js` are a good fit to prevent regressions.
